aeron-driver/src/main/c/aeron_driver_conductor.c

```diff
@@ -4183,83 +4579,114 @@ int aeron_driver_conductor_on_add_send_destination(
-                command->channel_length, command_uri);
-            goto error_cleanup;
-        }
+    size_t uri_length = (size_t)command->channel_length;
+    if (aeron_uri_parse(uri_length, command_uri, uri) < 0 ||
+        aeron_driver_conductor_validate_send_destination_uri(uri, uri_length) < 0)
+    {
+        goto error_cleanup;
+    }
 
-        if (uri->type != AERON_URI_UDP || NULL == uri->params.udp.endpoint)
-        {
-            AERON_SET_ERR(EINVAL, "incorrect URI format for destination: %.*s", command->channel_length, command_uri);
-            goto error_cleanup;
-        }
+    if (aeron_driver_conductor_validate_destination_uri_params(uri, uri_length) < 0)
+    {
+        AERON_APPEND_ERR("%s", "");
+        goto error_cleanup;
+    }
 
-        struct sockaddr_storage destination_addr;
-        if (aeron_name_resolver_resolve_host_and_port(
-            &conductor->name_resolver,
-            uri->params.udp.endpoint,
-            AERON_UDP_CHANNEL_ENDPOINT_KEY,
-            false,
-            &destination_addr) < 0)
-        {
-            memset(&destination_addr, 0, sizeof(destination_addr));
-            destination_addr.ss_family = AF_UNSPEC;
-        }
+    if (NULL == endpoint->destination_tracker || !endpoint->destination_tracker->is_manual_control_mode)
+    {
+        AERON_SET_ERR(
+            EINVAL,
+            "channel does not allow manual control of destinations: %.*s",
+            command->channel_length, command_uri);
+        goto error_cleanup;
+    }
 
-        aeron_driver_sender_proxy_on_add_destination(
-            conductor->context->sender_proxy, endpoint, uri, &destination_addr);
-        aeron_driver_conductor_on_operation_succeeded(conductor, command->correlated.correlation_id);
+    if (uri->type != AERON_URI_UDP || NULL == uri->params.udp.endpoint)
+    {
+        AERON_SET_ERR(EINVAL, "incorrect URI format for destination: %.*s", command->channel_length, command_uri);
+        goto error_cleanup;
+    }
 
-        return 0;
+    aeron_driver_async_client_command_t *async_client_command;
 
-error_cleanup:
-        aeron_uri_close(uri);
-        aeron_free(uri);
-        return -1;
+    if (aeron_driver_async_client_command_allocate(
+        &async_client_command,
+        command,
+        sizeof(aeron_destination_command_t)) < 0)
+    {
+        AERON_APPEND_ERR("%s", "");
+        goto error_cleanup;
     }
 
-    AERON_SET_ERR(
-        -AERON_ERROR_CODE_UNKNOWN_PUBLICATION,
-        "unknown add destination client_id=%" PRId64 " registration_id=%" PRId64,
-        command->correlated.client_id,
-        command->registration_id);
+    async_client_command->async_command.async_resolve.uri_param_name = AERON_UDP_CHANNEL_ENDPOINT_KEY;
+    async_client_command->async_command.async_resolve.is_re_resolution = false;
+    memcpy(async_client_command->async_command.async_resolve.endpoint_name, uri->params.udp.endpoint, strlen(uri->params.udp.endpoint));
```

Comment:
I think we should use `strnlen(..., AERON_MAX_HOST_LEN)` and null the final byte just for belt and braces.

Reply:
Should I define the field in the async_resolve struct to be AERON_MAX_HOST_LENGTH + 1?  Otherwise I can really only memcpy up to AERON_MAX_HOST_LEN - 1 so I can add the terminating null.